crates/accelerate/src/gate_direction.rs

```diff
@@ -142,8 +157,505 @@ where
+fn replace_dag(
+    py: Python,
+    std_gate: StandardGate,
+    inst: &PackedInstruction,
+) -> PyResult<DAGCircuit> {
+    let replacement_dag = match std_gate {
+        StandardGate::CXGate => cx_replacement_dag(py),
+        StandardGate::ECRGate => ecr_replacement_dag(py),
+        StandardGate::CZGate => cz_replacement_dag(py),
+        StandardGate::SwapGate => swap_replacement_dag(py),
+        StandardGate::RXXGate => rxx_replacement_dag(py, inst.params_view()),
+        StandardGate::RYYGate => ryy_replacement_dag(py, inst.params_view()),
+        StandardGate::RZZGate => rzz_replacement_dag(py, inst.params_view()),
+        StandardGate::RZXGate => rzx_replacement_dag(py, inst.params_view()),
+        _ => panic!("Mismatch in supported gates assumption"),
+    };
+
+    replacement_dag
+}
+
+//###################################################
+// Utility functions to build the replacement dags
+// TODO: replace once we have fully Rust-friendly versions of QuantumRegister, DAGCircuit and ParemeterExpression
+#[inline]
+fn create_qreg(py: Python<'_>, size: u32) -> PyResult<Bound<'_, PyAny>> {
+    imports::QUANTUM_REGISTER.get_bound(py).call1((size,))
+}
+
+#[inline]
+fn qreg_bit<'py>(py: Python, qreg: &Bound<'py, PyAny>, index: u32) -> PyResult<Bound<'py, PyAny>> {
+    qreg.call_method1(intern!(py, "__getitem__"), (index,))
+}
+
+#[inline]
+fn std_gate(py: Python, gate: StandardGate) -> PyResult<Py<PyAny>> {
+    gate.create_py_op(
+        py,
+        None,
+        &ExtraInstructionAttributes::new(None, None, None, None),
+    )
+}
+
+#[inline]
+fn parameterized_std_gate(py: Python, gate: StandardGate, param: &[Param]) -> PyResult<Py<PyAny>> {
+    debug_assert!(param.len() == 1); //
+    gate.create_py_op(
+        py,
+        Some(param),
+        &ExtraInstructionAttributes::new(None, None, None, None),
+    )
+}
+
+#[inline]
+fn apply_op_back(
+    py: Python,
+    dag: &mut DAGCircuit,
+    op: &Py<PyAny>,
+    qargs: &Vec<&Bound<PyAny>>,
+) -> PyResult<()> {
+    dag.py_apply_operation_back(
+        py,
+        op.bind(py).clone(),
+        Some(TupleLikeArg::extract_bound(&PyTuple::new_bound(py, qargs))?),
+        None,
+        false,
+    )?;
+
+    Ok(())
+}
+
+fn cx_replacement_dag(py: Python) -> PyResult<DAGCircuit> {
+    let qreg = create_qreg(py, 2)?;
+    let new_dag = &mut DAGCircuit::new(py)?;
+    new_dag.add_qreg(py, &qreg)?;
+
+    let (q0, q1) = (qreg_bit(py, &qreg, 0)?, qreg_bit(py, &qreg, 1)?);
+    apply_op_back(py, new_dag, &std_gate(py, StandardGate::HGate)?, &vec![&q0])?;
+    apply_op_back(py, new_dag, &std_gate(py, StandardGate::HGate)?, &vec![&q1])?;
+    apply_op_back(
+        py,
+        new_dag,
+        &std_gate(py, StandardGate::CXGate)?,
+        &vec![&q1, &q0],
+    )?;
+    apply_op_back(py, new_dag, &std_gate(py, StandardGate::HGate)?, &vec![&q0])?;
+    apply_op_back(py, new_dag, &std_gate(py, StandardGate::HGate)?, &vec![&q1])?;
+
+    Ok(new_dag.clone())
+}
```

Comment:
This and the subsequent `_replacement_dag()` methods could use more rust-natiuve approaches here without so much Python abstraction. I understand this was made before the methods existed, but now that we have them available, we should use them over the python exposed ones (e.g. use `DAGCircuit::apply_operation_back()` over `DAGCircuit::py_apply_operation_back()` and so on...).

Reply:
Done